File: app/schemas/usuarios.py
# app/schemas/usuarios.py
"""
Schemas Pydantic v2 para el dominio usuarios/empleados.
"""

from datetime import date, datetime
from typing import Optional

from pydantic import BaseModel

from app.schemas.contratos import ContratoEmpleadoResumen
from app.schemas.empleados import (
    AreaResponse,
    CategoriaResponse,
    ClasificacionEmpleadoResponse,
    EstadoEmpleadoResponse,
    PuestoResponse,
    SubareaResponse,
)


class UsuarioAsignacionUpdate(BaseModel):
    """Solo RH puede usar este schema. Permite cambiar lider_id, rol_id y comedor en turnos."""

    model_config = {"str_strip_whitespace": True}
    lider_id: Optional[int] = None
    rol_id: Optional[int] = None
    comedor_id: Optional[int] = None


class RolBrief(BaseModel):
    model_config = {"from_attributes": True}
    id: int
    nombre: str


class UsuarioResponse(BaseModel):
    model_config = {"from_attributes": True}

    id: int
    empleado_id: int
    no_empleado: int
    nombre: str
    email: Optional[str] = None
    rol_id: Optional[int] = None
    rol: Optional[RolBrief] = None
    estado: Optional[EstadoEmpleadoResponse] = None
    area: Optional[AreaResponse] = None
    subarea: Optional[SubareaResponse] = None
    puesto: Optional[PuestoResponse] = None
    categoria: Optional[CategoriaResponse] = None
    clasificacion: Optional[ClasificacionEmpleadoResponse] = None
    lider_id: Optional[int] = None
    centrocosto_id: Optional[int] = None
    foto: Optional[str] = None
    registro: Optional[str] = None
    created_at: Optional[datetime] = None


class UsuarioListItem(UsuarioResponse):
    """Fila de listado RH con nombre del líder resuelto."""

    lider_nombre: Optional[str] = None


class UsuarioPageResponse(BaseModel):
    items: list[UsuarioListItem]
    total: int
    page: int
    page_size: int


class EmpleadoDistribucionItem(BaseModel):
    """Conteo agrupado para gráficas del dashboard RH (resumen de plantilla)."""

    label: str
    total: int


class EmpleadosPorClasificacionAreaSerie(BaseModel):
    """Activos por área para una clasificación (catálogo `clasificacion_empleado`)."""

    tipo: str
    clasificacion_id: int | None = None
    clasificacion_descripcion: str
    por_area: list[EmpleadoDistribucionItem] = []


class UsuarioResumenResponse(BaseModel):
    total_plantilla: int
    activos: int
    inactivos: int
    sin_lider_asignado: int
    sin_email_administrativo: int = 0
    practicantes: int
    porcentaje_operatividad: float
    colaboradores_total: int
    contratos_por_vencer: int
    empleados_por_clasificacion_y_area: list[EmpleadosPorClasificacionAreaSerie] = []


class CatalogoFiltrosResponse(BaseModel):
    areas: list[AreaResponse]
    puestos: list[PuestoResponse]


class SolicitudBrief(BaseModel):
    model_config = {"from_attributes": True}
    id: int
    tipo: str
    estado: str
    fecha_inicio: date
    fecha_fin: date
    created_at: datetime


class IncidenciaBrief(BaseModel):
    model_config = {"from_attributes": True}
    id: int
    tipo: str
    estatus_id: Optional[int] = None
    created_at: datetime


class ActaBrief(BaseModel):
    model_config = {"from_attributes": True}
    id: int
    estado: str
    created_at: datetime


class Vista360TurnoEmpleado(BaseModel):
    """Filas de `turnos_empleados` (nombre de comedor + turno). Solo se envía cuando el solicitante es RH."""

    comedor: Optional[str] = None
    turno: Optional[str] = None


class UsuarioVista360Response(BaseModel):
    model_config = {"from_attributes": True}
    usuario: UsuarioResponse
    solicitudes_recientes: list[SolicitudBrief]
    incidencias_activas: list[IncidenciaBrief]
    actas_firmadas: list[ActaBrief]
    turno_empleado: Optional[Vista360TurnoEmpleado] = None
    # Fecha de ingreso real desde SQL Server datos-analisis (CB_FEC_ING de dbo.COLABORA);
    # None si la BD externa no está disponible (el frontend cae a `registro`).
    fecha_ingreso: Optional[date] = None
    # Contrato actual (misma caché). None si el empleado no está sincronizado.
    contrato: Optional[ContratoEmpleadoResumen] = None


class MetricasUsuarioResponse(BaseModel):
    solicitudes_por_estado: dict[str, int]
    incidencias_por_tipo: dict[str, int]
    dias_antiguedad: int
    total_actas: int

File: frontend/src/api/usuarios.ts
/** Tipos compartidos (directorio en GET /api/v1/empleados; CRUD en /api/v1/usuarios). */

export type EmpleadoDistribucionItem = {
  label: string;
  total: number;
};

export type EmpleadosClasificacionTipo = "administrativo" | "directo" | "indirecto";

export type EmpleadosPorClasificacionAreaSerie = {
  tipo: EmpleadosClasificacionTipo;
  clasificacion_id: number | null;
  clasificacion_descripcion: string;
  por_area: readonly EmpleadoDistribucionItem[];
};

export type UsuarioResumen = {
  total_plantilla: number;
  activos: number;
  inactivos: number;
  sin_lider_asignado: number;
  /** Administrativos activos sin correo registrado (empleado.email ni tabla emails). */
  sin_email_administrativo: number;
  practicantes: number;
  porcentaje_operatividad: number;
  /** Activos en el alcance del rol (plantilla completa para RH; equipo para supervisor/gerente). */
  colaboradores_total: number;
  /** Contratos con fin en los próximos 30 días (mismo alcance). */
  contratos_por_vencer: number;
  /** Activos por área, desglosados por clasificación (solo resumen RH / plantilla completa). */
  empleados_por_clasificacion_y_area?: readonly EmpleadosPorClasificacionAreaSerie[];
};

export type RolBrief = { id: number; nombre: string };

export type CatalogoBase = { descripcion: string; estatus_id: number };

export type AreaResponse = CatalogoBase & { area_id: number };

export type PuestoResponse = CatalogoBase & { puesto_id: number; area_id?: number | null };

export type SubareaResponse = CatalogoBase & { subarea_id: number; area_id: number };

export type EstadoEmpleadoResponse = CatalogoBase & { estado_id: number };

export type CategoriaResponse = {
  categoria_id: number;
  nivel?: string | null;
  descripcion?: string | null;
  estatus_id: number;
};

export type ClasificacionEmpleadoResponse = CatalogoBase & {
  clasificacion_id: number;
  significado?: string | null;
};

/** Fila de listado / UsuarioResponse del API. */
export type UsuarioListItem = {
  id: number;
  empleado_id: number;
  no_empleado: string;
  nombre: string;
  email: string | null;
  rol_id: number;
  rol: RolBrief | null;
  estado: EstadoEmpleadoResponse | null;
  area: AreaResponse | null;
  subarea: SubareaResponse | null;
  puesto: PuestoResponse | null;
  categoria: CategoriaResponse | null;
  clasificacion: ClasificacionEmpleadoResponse | null;
  lider_id: number | null;
  /** URL o ruta de foto de perfil (listado directorio). */
  foto?: string | null;
  /** Solo en GET /empleados (listado). */
  lider_nombre?: string | null;
  registro: string | null;
  created_at: string;
};

export type UsuarioPage = {
  items: UsuarioListItem[];
  total: number;
  page: number;
  page_size: number;
};

export type CatalogoFiltros = {
  areas: AreaResponse[];
  puestos: PuestoResponse[];
};

export type UsuariosFetchError = {
  status: number;
  detail: string;
};

export function isUsuariosFetchError(e: unknown): e is UsuariosFetchError {
  return (
    typeof e === "object" &&
    e !== null &&
    "status" in e &&
    "detail" in e &&
    typeof (e as UsuariosFetchError).detail === "string"
  );
}
